pkg/controller/deployment/deployment_controller.go

```diff
@@ -450,6 +451,11 @@ func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extension
 		return dc.updateDeploymentStatus(allRCs, newRC, deployment)
 	}
 
+	// Cleanup old rcs
+	if deployment.Spec.RevisionHistoryLimit != nil {
```

Comment:
You mean implement like this:

```
switch d.Spec.Strategy.Type {
	case extensions.RecreateDeploymentStrategyType:
		if err:= dc.syncRecreateDeployment(d); err != nil {
			return err
		}
	case extensions.RollingUpdateDeploymentStrategyType:
		if err:= dc.syncRollingUpdateDeployment(d); err != nil {
			return err
		}
	}
}

// Cleanup old rcs
if deployment.Spec.RevisionHistoryLimit != nil {
	return dc.cleanupOldRcs(deployment)
}
```

I think it's not viable, since we can only delete old RCs after the RCs scaled down to 0, but `syncRollingUpdateDeployment` can return before scale down, see https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L464